### PR TITLE
gh-135836: Fix IndexError in asyncio.create_connection with empty exceptions list

### PR DESCRIPTION
**Pull request by heliang666s**
Fix IndexError that occurs in asyncio.create_connection() when Happy Eyeballs algorithm leaves empty sublists in the exceptions list, which after flattening becomes an empty list causing str(exceptions[0]) to crash. The fix adds explicit handling for empty exceptions list by changing 'else:' to 'elif len(exceptions) > 1:' and adding a new else clause that raises OSError('create_connection failed') instead of crashing.

* Issue: gh-135836

### CODE REVIEW

File: Lib/asyncio/base_events.py [hunk 1]
```diff
@@ -1170,6 +1170,9 @@ async def create_connection(
                         # the various error messages.
                         raise OSError('Multiple exceptions: {}'.format(
                             ', '.join(str(exc) for exc in exceptions)))
+                    else:
+                        # No exceptions were collected, raise a generic connection error
+                        raise OSError('create_connection failed')
```
Comment (serhiy-storchaka) [inline Lib/asyncio/base_events.py]: Would not be better to raise a `TimeoutError`? This only happens when all connections exceed their time.

**Review by serhiy-storchaka — COMMENTED** [review]
The approach LGTM, but please add a test. You can take `test_create_connection_timeout` as example. It only tests for `happy_eyeballs_delay=None` (default), we need a test for non-None value.

File: Misc/NEWS.d/next/Library/2025-06-23-11-04-25.gh-issue-135836.-C-c4v.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Fixed :exc:`IndexError` in ``asyncio.create_connection()`` that could occur when the Happy Eyeballs algorithm resulted in an empty exceptions list during connection attempts.
```
Comment (ZeroIntensity) [inline Misc/NEWS.d/next/Library/2025-06-23-11-04-25.gh-issue-135836.-C-c4v.rst]: Active voice and use a Sphinx ref:
```suggestion
Fix :exc:`IndexError` in :meth:`asyncio.loop.create_connection` that could occur when the Happy Eyeballs algorithm resulted in an empty exceptions list during connection attempts.
```

**Comment by serhiy-storchaka** [comment]
Please don't forget about test.

**Comment by heliang666s** [comment]
I think this pr is ready for merge. cc @serhiy-storchaka @ZeroIntensity

File: Lib/test/test_asyncio/test_base_events.py [hunk 3]
```diff
@@ -1190,6 +1190,37 @@ def getaddrinfo(*args, **kw):
...
+            # Return (None, []) - no winner, empty exceptions list
+            async def mock_race(coro_fns, delay, loop):
+                return None, []
+            mock_staggered.side_effect = mock_race
+
+            coro = self.loop.create_connection(
+                MyProto, 'example.com', 80, happy_eyeballs_delay=0.1)
+
+            # Should raise TimeoutError instead of IndexError
+            with self.assertRaises(TimeoutError):
```
Comment (kumaraditya303) [inline Lib/test/test_asyncio/test_base_events.py]: Can you add check for the error message as well?

File: Lib/test/test_asyncio/test_base_events.py [hunk 4]
```diff
@@ -1190,9 +1190,44 @@ def getaddrinfo(*args, **kw):
...
+            # Return (None, []) - no winner, empty exceptions list
+            async def mock_race(coro_fns, delay, loop):
+                return None, []
+            mock_staggered.side_effect = mock_race
+
+            coro = self.loop.create_connection(
+                MyProto, 'example.com', 80, happy_eyeballs_delay=0.1)
+
+            # Should raise TimeoutError instead of IndexError
+            with self.assertRaisesRegex(TimeoutError, "connection timed out"):
```
Comment (kumaraditya303) [inline Lib/test/test_asyncio/test_base_events.py]: The msg should be "create_connection failed" no?

File: Lib/test/test_asyncio/test_base_events.py [hunk 5]
```diff
@@ -1190,9 +1190,44 @@ def getaddrinfo(*args, **kw):
...
+
+            # Should raise TimeoutError instead of IndexError
+            with self.assertRaisesRegex(TimeoutError, "create_connection failed"):
+                self.loop.run_until_complete(coro)
+
     def test_create_connection_host_port_sock(self):
+        # host, port and sock are specified
         coro = self.loop.create_connection(
-            MyProto, 'example.com', 80, sock=object())
+            MyProto, 'example.com', 80, sock=mock.Mock())
```
Comment (kumaraditya303) [inline Lib/test/test_asyncio/test_base_events.py]: Why this change is needed?

File: Lib/asyncio/base_events.py [hunk 6]
```diff
@@ -1161,7 +1161,7 @@ async def create_connection(
                         raise ExceptionGroup("create_connection failed", exceptions)
                     if len(exceptions) == 1:
                         raise exceptions[0]
-                    else:
+                    elif len(exceptions) > 1:
```
Comment (kumaraditya303) [inline Lib/asyncio/base_events.py]:
```suggestion
                    elif exceptions:
```

File: Lib/test/test_asyncio/test_base_events.py [hunk 7]
```diff
@@ -1190,6 +1190,36 @@ def getaddrinfo(*args, **kw):
                 self.loop.run_until_complete(coro)
             self.assertTrue(sock.close.called)
 
+    @patch_socket
+    def test_create_connection_happy_eyeballs_empty_exceptions(self, m_socket):
+        # Test for gh-135836: Fix IndexError when Happy Eyeballs algorithm
+        # results in empty exceptions list
```
Comment (kumaraditya303) [inline Lib/test/test_asyncio/test_base_events.py]:
```suggestion
        # See gh-135836: Fix IndexError when Happy Eyeballs algorithm
        # results in empty exceptions list
```

**Review by kumaraditya303 — APPROVED** [review]
